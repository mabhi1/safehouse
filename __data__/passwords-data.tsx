export const passwordWithoutError = {
  data: [
    {
      id: "id1",
      createdAt: new Date("2024-09-10T05:34:42.367Z"),
      password: "U2FsdGVkX19F9RWkBzfFGod1yJPlLNMvBJR8n3kwp3c=",
      site: "site1",
      uid: "uid1",
      updatedAt: new Date("2024-09-10T05:34:42.367Z"),
      username: "username1",
    },
    {
      id: "id2",
      createdAt: new Date("2024-09-10T05:34:42.367Z"),
      password: "U2FsdGVkX19F9RWkBzfFGod1yJPlLNMvBJR8n3kwp3c=",
      site: "site2",
      uid: "uid2",
      updatedAt: new Date("2024-09-10T05:34:42.367Z"),
      username: "username2",
    },
  ],
  error: null,
};

export const passwordWithError = {
  data: null,
  error: true,
};
