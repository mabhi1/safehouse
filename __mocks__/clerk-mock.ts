import { auth } from "@clerk/nextjs/server";

const mockAuth = auth as jest.Mock;
jest.mock("@clerk/nextjs/server", () => ({
  auth: jest.fn(),
}));

export { mockAuth };
