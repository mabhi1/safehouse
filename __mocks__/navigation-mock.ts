jest.mock("next/navigation", () => ({
  useSearchParams: jest.fn(() => ({
    get: jest.fn(() => "lastUpdated"),
  })),
  useRouter: jest.fn(() => ({
    push: jest.fn(),
  })),
  usePathname: jest.fn(() => "/"),
}));
