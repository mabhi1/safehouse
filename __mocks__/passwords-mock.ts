const mockGetPasswordsByUser = jest.fn();
jest.mock("../prisma/db/passwords", () => ({
  getPasswordsByUser: () => mockGetPasswordsByUser(),
}));

const mockAddPassword = jest.fn();
const mockEditPassword = jest.fn();
jest.mock("../actions/passwords", () => ({
  addPassword: () => mockAddPassword(),
  editPassword: () => mockEditPassword(),
}));

export { mockGetPasswordsByUser, mockAddPassword, mockEditPassword };
