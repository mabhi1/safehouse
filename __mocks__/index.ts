import "./navigation-mock";
import "./clerk-mock";
import "./passwords-mock";
