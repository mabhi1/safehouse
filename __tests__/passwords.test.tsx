import "@testing-library/jest-dom";
import "@/__mocks__";
import { passwordWithoutError } from "@/__data__/passwords-data";
import Passwords from "@/app/(web)/passwords/page";
import { render, screen } from "@testing-library/react";
import { mockGetPasswordsByUser } from "@/__mocks__/passwords-mock";
import { mockAuth } from "@/__mocks__/clerk-mock";

beforeAll(() => {
  mockGetPasswordsByUser.mockResolvedValue(passwordWithoutError);
  mockAuth.mockImplementation(() => ({ userId: passwordWithoutError.data[0].id }));
});

describe("Password Page", () => {
  it("renders page with empty passwords data", async () => {
    const { container } = render(await Passwords({ searchParams: { search: "" } }));
    expect(container).toMatchSnapshot();
    expect(screen.getByText("Passwords")).toBeInTheDocument();
  });

  it("renders page with passwords data and empty search string", async () => {
    render(await Passwords({ searchParams: { search: "" } }));
    const passwords = screen.getAllByTestId(/password/);
    expect(passwords).toHaveLength(2);
  });

  it("renders page with matching site", async () => {
    render(await Passwords({ searchParams: { search: passwordWithoutError.data[0].site } }));
    const passwords = screen.getAllByTestId(/password/);
    expect(passwords).toHaveLength(1);
  });

  it("renders page with matching username", async () => {
    render(await Passwords({ searchParams: { search: passwordWithoutError.data[1].username } }));
    const passwords = screen.getAllByTestId(/password/);
    expect(passwords).toHaveLength(1);
  });

  it("navigates to signin page with no user id", async () => {
    const mockRedirectToSignIn = jest.fn();
    mockAuth.mockImplementationOnce(() => ({ userId: null, redirectToSignIn: () => mockRedirectToSignIn() }));
    render(await Passwords({ searchParams: { search: "" } }));
    expect(mockRedirectToSignIn).toHaveBeenCalled();
  });

  it("renders error page with invalid passwords data", async () => {
    mockGetPasswordsByUser.mockResolvedValue({
      data: null,
      error: true,
    });
    try {
      render(await Passwords({ searchParams: { search: "" } }));
    } catch (error) {
      expect(error).toStrictEqual(new Error("User not found"));
    }
  });
});
