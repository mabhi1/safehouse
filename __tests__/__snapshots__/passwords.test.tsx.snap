// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password Page renders page with empty passwords data 1`] = `
<div>
  <div
    class="space-y-5"
  >
    <div
      class="flex items-center gap-3 md:gap-5"
    >
      <div
        class="flex items-center mr-auto gap-1"
      >
        <span
          class="text-base uppercase"
        >
          Passwords
        </span>
        <div
          class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 font-normal"
        >
          2
        </div>
      </div>
      <button
        aria-autocomplete="none"
        aria-controls="radix-:r0:"
        aria-expanded="false"
        class="flex h-10 items-center justify-between rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-52"
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          data-testid="sortPasswordsSelect"
          style="pointer-events: none;"
        >
          Last Updated
        </span>
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down h-4 w-4 opacity-50"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
      <div
        aria-controls="radix-:r1:"
        aria-expanded="false"
        aria-haspopup="dialog"
        data-state="closed"
        type="button"
      >
        <button
          class="items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 hidden md:block"
          data-testid="addPasswordButton"
        >
          Add Password
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 w-9 md:hidden"
        >
          <svg
            class="lucide lucide-plus w-[1.2rem] h-[1.2rem]"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
            <path
              d="M12 5v14"
            />
          </svg>
        </button>
      </div>
    </div>
    <ul
      class="grid grid-cols-1 md:grid-cols-3 gap-5 w-full"
    >
      <li
        class="flex flex-col justify-between shadow transition-shadow duration-300 hover:shadow-lg border rounded p-2 overflow-hidden"
        data-testid="passwordid1"
      >
        <div
          class="flex flex-col gap-2"
        >
          <div
            class="break-words"
          >
            site1
          </div>
          <div
            class="break-words"
          >
            username1
          </div>
          <div
            class="cursor-pointer w-full truncate"
            data-testid="togglePassword"
          >
            U2FsdGVkX19F9RWkBzfFGod1yJPlLNMvBJR8n3kwp3c=
          </div>
        </div>
        <div>
          <div
            class="flex justify-end my-1"
          >
            <button
              aria-controls="radix-:r4:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2"
              data-state="closed"
              data-testid="editPasswordButton"
              type="button"
            >
              Edit
            </button>
            <button
              aria-controls="radix-:r7:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-accent h-9 px-4 py-2 text-destructive hover:text-destructive"
              data-state="closed"
              type="button"
            >
              Delete
            </button>
          </div>
          <div
            class="text-slate-500 text-xs border-t border-slate-500 pt-1"
          >
            Updated : 
            September 10, 2024
          </div>
        </div>
      </li>
      <li
        class="flex flex-col justify-between shadow transition-shadow duration-300 hover:shadow-lg border rounded p-2 overflow-hidden"
        data-testid="passwordid2"
      >
        <div
          class="flex flex-col gap-2"
        >
          <div
            class="break-words"
          >
            site2
          </div>
          <div
            class="break-words"
          >
            username2
          </div>
          <div
            class="cursor-pointer w-full truncate"
            data-testid="togglePassword"
          >
            U2FsdGVkX19F9RWkBzfFGod1yJPlLNMvBJR8n3kwp3c=
          </div>
        </div>
        <div>
          <div
            class="flex justify-end my-1"
          >
            <button
              aria-controls="radix-:ra:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2"
              data-state="closed"
              data-testid="editPasswordButton"
              type="button"
            >
              Edit
            </button>
            <button
              aria-controls="radix-:rd:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-accent h-9 px-4 py-2 text-destructive hover:text-destructive"
              data-state="closed"
              type="button"
            >
              Delete
            </button>
          </div>
          <div
            class="text-slate-500 text-xs border-t border-slate-500 pt-1"
          >
            Updated : 
            September 10, 2024
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
